src/main/java/org/broadinstitute/hellbender/engine/dataflow/transforms/composite/AddContextDataToReadOptimizedFn.java

```diff
@@ -0,0 +1,280 @@
+import org.apache.spark.api.java.function.FlatMapFunction;
+import org.broadinstitute.hellbender.engine.ReadContextData;
+import org.broadinstitute.hellbender.engine.ContextShard;
+import org.broadinstitute.hellbender.engine.datasources.ReferenceMultiSource;
+import org.broadinstitute.hellbender.engine.filters.ReadFilter;
+import org.broadinstitute.hellbender.engine.spark.AuthHolder;
+import org.broadinstitute.hellbender.exceptions.GATKException;
+import org.broadinstitute.hellbender.utils.IntervalUtils;
+import org.broadinstitute.hellbender.utils.SimpleInterval;
+import org.broadinstitute.hellbender.utils.collections.IntervalsSkipList;
+import org.broadinstitute.hellbender.utils.gcs.BucketUtils;
+import org.broadinstitute.hellbender.utils.read.GATKRead;
+import org.broadinstitute.hellbender.utils.read.SAMRecordToGATKReadAdapter;
+import org.broadinstitute.hellbender.utils.reference.ReferenceBases;
+import org.broadinstitute.hellbender.utils.variant.Variant;
+
+
+import java.io.File;
+import java.io.IOException;
+import java.io.Serializable;
+import java.security.GeneralSecurityException;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.NoSuchElementException;
+
+
+public final class AddContextDataToReadOptimizedFn implements Serializable{
+    private static final long serialVersionUID = 1L;
+
+    /**
+     * Fill in reads that start in the given shard, and subshard the output to the requested size.
+     * Signals an error if any read sticks out more than "margin" outside of the big shard.
+     * Optionally filters the reads to only keep the ones that satisfy the given predicate.
+     *
+     * Hadoop paths aren't yet supported.
+     *
+     *   _   _  ____ _______ ______
+     *  | \ | |/ __ \__   __|  ____|
+     *  |  \| | |  | | | |  | |__
+     *  | . ` | |  | | | |  |  __|
+     *  | |\  | |__| | | |  | |____
+     *  |_| \_|\____/  |_|  |______| (important):
+     *
+     * The reads are stored without the header.
+     *
+     * @param bam URL to the reads
+     * @param outputShardSize subsharding size
+     * @param margin throw an exception if any read sticks out beyond the margin
+     * @param optFilter if specified, only reads that satisfy this will be included.
+     * @return A DoFn that acts as described above.
+     * @throws IOException
+     */
+    public static FlatMapFunction<ContextShard,ContextShard> subdivideAndFillReads(String bam, AuthHolder auth, int outputShardSize, int margin, final ReadFilter optFilter) throws IOException {
+            return new FlatMapFunction<ContextShard, ContextShard>() {
+                private static final long serialVersionUID = 1L;
+                @Override
+                public Iterable<ContextShard> call(ContextShard contextShard) throws Exception {
+                    final SubdivideAndFillReadsIterator iterator = new SubdivideAndFillReadsIterator(bam, auth, outputShardSize, margin, optFilter, contextShard);
```

Comment:
Why are you creating this apparently-unused local iterator? Isn't an identical iterator returned from `iterator()` below?